processing/src/main/java/io/druid/segment/data/CompressedVSizeIntsIndexedSupplier.java

```diff
@@ -0,0 +1,393 @@
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package io.druid.segment.data;
+
+import com.google.common.base.Preconditions;
+import com.google.common.io.Closeables;
+import com.google.common.primitives.Ints;
+import com.google.common.primitives.Shorts;
+import com.metamx.common.IAE;
+import com.metamx.common.guava.CloseQuietly;
+import io.druid.collections.ResourceHolder;
+import io.druid.collections.StupidResourceHolder;
+import io.druid.segment.CompressedPools;
+
+import java.io.IOException;
+import java.nio.ByteBuffer;
+import java.nio.ByteOrder;
+import java.nio.IntBuffer;
+import java.nio.ShortBuffer;
+import java.nio.channels.WritableByteChannel;
+import java.util.Iterator;
+import java.util.List;
+
+public class CompressedVSizeIntsIndexedSupplier implements WritableSupplier<IndexedInts>
+{
+  public static final byte version = 0x2;
+
+  private final int totalSize;
+  private final int sizePer;
+  private final int numBytes;
+  private final int bigEndianShift;
+  private final int littleEndianMask;
+  private final GenericIndexed<ResourceHolder<ByteBuffer>> baseBuffers;
+  private final CompressedObjectStrategy.CompressionStrategy compression;
+
+  CompressedVSizeIntsIndexedSupplier(
+      int totalSize,
+      int sizePer,
+      int numBytes,
+      GenericIndexed<ResourceHolder<ByteBuffer>> baseBuffers,
+      CompressedObjectStrategy.CompressionStrategy compression
+  )
+  {
+    Preconditions.checkArgument(
+        sizePer == (1 << Integer.numberOfTrailingZeros(sizePer)),
+        "Number of entries per chunk must be a power of 2"
+    );
+
+    this.totalSize = totalSize;
+    this.sizePer = sizePer;
+    this.baseBuffers = baseBuffers;
+    this.compression = compression;
+    this.numBytes = numBytes;
+    this.bigEndianShift = Integer.SIZE - (numBytes << 3); // numBytes * 8
+    this.littleEndianMask = (int)((1L << (numBytes << 3)) - 1); // set numBytes * 8 lower bits to 1
+  }
+
+  public static int maxIntsInBufferForBytes(int numBytes)
+  {
+    int maxSizePer = (CompressedPools.BUFFER_SIZE - bufferPadding(numBytes)) / numBytes;
+    // round down to the nearest power of 2
+    return 1 << (Integer.SIZE - 1 - Integer.numberOfLeadingZeros(maxSizePer));
+  }
+
+  private static int bufferPadding(int numBytes)
+  {
+    // when numBytes == 3 we need to pad the buffer to allow reading an extra byte
+    // beyond the end of the last value, since we use buffer.getInt() to read values.
+    // for numBytes < 3 we remove the need for padding by reading short or bytes directly.
+    if (numBytes == 3) {
```

Comment:
the idea of variable sized columns is that we store positive integers using either 1, 2, 3, or 4 bytes.
Because we use sometimes use getInt with some bit magic to read the actual value, we have to add padding at the end of the buffer in order to always be able to read 4 bytes.

For 1, 2, and 4 byte ints, we use get(), getShort(), and getInt() respectively, so we don't need to pad. For 3-byte ints, we have to add 1 byte padding.